_posts/2021-11-02-choosing-dask-chunk-sizes.md

```diff
@@ -0,0 +1,157 @@
+
+Array chunks can't be too big (we'll run out of memory), or too small (the overhead introduced by Dask becomes overwhelming). So how can we get it right?
+
+It's a two step process:
+
+1. First, start by choosing a chunk size similar to data you know can be processed entirely within memory (i.e. without dask)
+2. Then, watch the dask dashboard task stream and worker memory plots, and adjust if needed.
+
+
+## Contents
+
+- [ What are Dask array chunks?](#what-are-dask-array-chunks)
+- [Choosing an initial chunk size](#choosing-an-initial-chunk-size)
+- [Too small is a problem](#too-small-is-a-problemg)
+- [Too big is also a problem](#too-big-is-also-a-problem)
+- [Using the Dask dashboard](#using-the-Dask-dashboard)
+- [Unmanaged memory](#unmanaged-memory)
+- [Thanks for reading](#thanks-for-reading)
+
+## What are Dask array chunks?
+
+Dask arrays are big structures, made out of many small chunks.
+Typically, each small chunk is an individual [numpy array](https://numpy.org/), and they are arranged together to make a much larger Dask array.
+
+<img src="https://raw.githubusercontent.com/dask/dask/ac01ddc9074365e40d888f80f5bcd955ba01e872/docs/source/images/dask-array-black-text.svg" alt="Diagram: Dask array chunks" width="400" height="300" />
+
+You can find more information about Dask array chunks on this page of the documentation: https://docs.dask.org/en/latest/array-chunks.html
+
+### How do I nkow what chunks my array has?
+
+If you have a Dask array, you can use the `chunksize` or `chunks` attribues to see information about the chunks. You can also visualize this with the Dask array HTML representation.
+
+![Visualizating Dask array chunks with the HTML repr](/images/choosing-good-chunk-sizes/examine-dask-array-chunks.png)
+
+`arr.chunksize` shows the largest chunk size. For arrays where you expect roughly uniform chunk sizes, this is a good way to summarize chunk size information.
+
+`arr.chunks` shows fully explicit sizes of all chunks along all dimensions within the Dask array (see [item 3 here](https://docs.dask.org/en/stable/array-chunks.html#specifying-chunk-shapes)). This is more verbose, and is a good choice with arrays that have irregular chunks.
+
+## Choosing an initial chunk size
+
+Some very rough rules of thumb:
+
+- When developing a new data processing pipeline, many people create a prototype using a small subset of the data, which may not involve Dask at all. This is helpful for quickly iterating on ideas, and also has the secondary benefit of giving you a clear idea of what size data can be processed easily for this workflow. You can use this knowledge to choose similar sized chunks in Dask.
+- Some people have observed that chunk sizes below 1MB are almost always bad. Chunk size between 50MB and 1GB are generally good (and chunk sizes over 2GB may be too big - but this upper limit varies a huge amount depending on the computational resources you have available, so adjust for your circumstances)
```

Comment:
The link given at the beginning of the article even say "it is rare to see chunk sizes below 100 MB". Not sure how to bring this here, or if we should...

Reply:
So perhaps this should say instead "chunk sizes between 100MB and 1GB are generally good"